Promote the `scep` package to root of repository

hslatman opened the pull request:
Example usage:

```golang
package main

import (
    "fmt"

    "github.com/smallstep/scep"
)

func main() {
    fmt.Println(scep.SUCCESS)
}
```

Example diff when migrating:

```console
-       github.com/micromdm/scep/v2 v2.1.0
+       github.com/smallstep/scep v0.0.0-20230420093139-1fcc424c7849
...
-       microscep "github.com/micromdm/scep/v2/scep"
+       smallscep "github.com/smallstep/scep"
```

Have done some quick sanity testing when importing this in a new project as well as in https://github.com/smallstep/certificates/ using `go get github.com/smallstep/scep@1fcc424c7849d4f705ab58e715b647c9e34dd074`. Results look OK.

Tests are also OK (still need to wire this repo into CI):
```console
$ go test ./...
ok  	github.com/smallstep/scep	0.899s
ok  	github.com/smallstep/scep/cryptoutil	0.288s
ok  	github.com/smallstep/scep/x509util	0.333s
```

~~I think the `v2` makes sense, and then we can continue with `v3` when we make a breaking change. This `v2` can be dropped in without big changes (just the `/scep` suffix to be removed, but they're already changing the path, so no issue, imo), so as a user of the package, it makes sense to me to import this as a `v2`.~~

We decided to start with a `v1` that's backwards compatible with the MicroMDM source. When we start breaking things, we'll do a `v2`.

hslatman committed:
Remove non-core files from the `scep` repo

hslatman committed:
Move `scep` package to root of the repository

hslatman committed:
Remove dependencies that are no longer required

hslatman committed:
Rename package to `github.com/smallstep/scep/v3`

This makes the package usable directly instead of using a replace
and signals that we're starting a new major version.

Moving from `github.com/micromdm/scep/v2` to relying on
this package requires the following changes:

 - The `scep` package is now top level, so can be imported directly
   as `github.com/smallstep/scep/v3`. This requires updating some
   import paths.
 - The `x509util` package was moved up one level.

hslatman committed:
Make it a `v2` instead

The changes so far should be backwards compatible with
`github.com/micromdm/scep/v2/scep`, so let's keep the `v2` for
now. We can bump it to `v3` when we make a (slightly) breaking
change, such as removing the `go-kit` leveled logging import.

azazeal commented on go.mod:
```diff
@@ -1,19 +1,12 @@
-module github.com/micromdm/scep/v2
+module github.com/smallstep/scep/v2
```
:thinking: should _we_ start with `v2`?

hslatman replied:
I did think about your remark about `v1` yesterday, but I think it makes sense, yes. The current functionality is backwards and drop-in compatible. I think this makes it simpler to explain to users that currently import `micromdm/scep/v2` that they can simply replace it with our `v2`. Then our next version, which I see happening very soon (removing `go-kit`), will be `v3`, which will then be a logical and spiritual successor to both our `v2` as well as MicroMDM's. I've seen this pattern before, for example with `go-jose`: moved to a different organization, but continued numbering.

maraino replied:
As far as I see, we are only moving `github.com/micromdm/scep/v2/scep`, I don't think there is a need for `v2` at the moment, the replacement will be quite similar, and I think it's more confusing than anything else. Especially because we won't have a `v1`.

We should also change this repo to be a source, not a fork. If not by default, GitHub will should you a massive merge when you do a PR on this repo.

azazeal reviewed (commented):
Let's also have @maraino go over this too

azazeal commented:
Cool beans are cool but there's that bit about the untyped strings that are
now typed. Thinking...

On Wed, Apr 19, 2023, 15:02 Herman Slatman ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In go.mod
> <https://github.com/smallstep/scep/pull/7#discussion_r1171237269>:
>
> > @@ -1,19 +1,12 @@
> -module github.com/micromdm/scep/v2
> +module github.com/smallstep/scep/v2
>
> I think it makes sense, yes. The current functionality is backwards and
> drop-in compatible. I think this makes it simpler to explain to users that
> currently import micromdm/scep/v2 that they can simply replace it with
> our v2. Then our next version, which I see happening very soon, will be v3,
> which will then be a logical and spiritual successor to both our v2 as
> well as MicroMDM's. I've seen this pattern before, for example with
> go-jose: moved to a different organization, but continued numbering.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/smallstep/scep/pull/7#discussion_r1171237269>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALAFSYYERHQQ6VMQLIEMTXB7H65ANCNFSM6AAAAAAXDZIQUM>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

hslatman commented:
> Cool beans are cool but there's that bit about the untyped strings that are now typed. Thinking...
> […](#)

Every first one of those was already typed: https://github.com/micromdm/scep/blob/main/scep/scep.go#L30-L46, I simply completed the rest, because the linter said so. 

I don't think this specific change makes the code backwards incompatible. If importers of this package put different strings in, then that's a programmer error, imo. I did a couple of quick checks of usage, and so far didn't find very problematic things.

maraino reviewed (commented):
As I commented before, this should be a source, and there's no point in using v2, as we won't have a v1. Let's do the v2, when the API changes.

hslatman committed:
Start at `v1`

hslatman committed:
Make `CreateCertificateRequest` public again

This exports `CreateCertificateRequest`, after it had been made
private. This allows using it in other projects again.
